src/base/dof_map_constraints.C

```diff
@@ -1406,17 +1406,20 @@ public:
     // are defined on the same System.
     System * system = nullptr;
 
-    // Map from boundary_id -> set<DirichletBoundary*> objects which
+    // Map from boundary_id -> set<pair<id,DirichletBoundary*>> objects which
     // are active on that boundary_id. Later we will use this to determine
     // which DirichletBoundary objects to loop over for each Elem.
-    std::map<boundary_id_type, std::set<DirichletBoundary *>>
-      boundary_id_to_dirichlet_boundaries;
+    std::map<boundary_id_type, std::set<std::pair<unsigned int, DirichletBoundary *>>>
+      boundary_id_to_ordered_dirichlet_boundaries;
 
-    for (const auto & dirichlet : dirichlets)
+    for (auto dirichlet_id : index_range(dirichlets))
```

Comment:
Unrelated to this PR, but I wonder if it would make sense to create an enumerate_range() function to iterate over both range and value at once, in a std::pair?  Aside from saving a line of code, that would enable optimizers to stop fretting about whether the vector was getting sneakily resized by any functions called from inside the loop.

Or would that be too much of a pain to use and/or to read (e.g. dirichlet.first and dirichlet.second being much less self-documenting than dirichlet_id and dirichlet) without C++17 structured bindings?

Reply:
Yeah I think your idea makes good sense once we require C++17 and structured bindings. IIRC there are several places in the library where we currently use an `index_range` but an `enumerate_range` is what we really want.